### ISSUE TITLE
Keras models with shared layers are loaded incorrectly from h5 file

### ISSUE THREAD

**Issue by hidanom** [tensorflow/tensorflow#50322]
**System information**
Default colab instance.

**Describe the current behavior**
Results of the different usages of a shared layer are permuted after the save/load procedure (h5 saved model format). This changes the network topology in an unpredictable manner.

**Describe the expected behavior**
Results are not permuted and the topology is intact.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no):  yes
- Briefly describe your candidate solution(if contributing):
The problem seems to lie in this part of model loading code (it has moved to a different file in tf 2.5, but is essentially the same)
https://github.com/tensorflow/tensorflow/blob/0931ea3d985bb9c8fdd054a5e29c4129623c849b/tensorflow/python/keras/engine/network.py#L1834-L1836
It handles the nodes we cannot compute right now. Such nodes are put back to the queue.
The problem arises when we process a shared layer (so there are several nodes associated with it) and some of the first nodes are not ready for computation, while some later nodes are. This leads to the order of nodes' computation being broken. This broken order is then used to find inputs for the following nodes, which leads to results permutation.

I propose two possible solutions:
1. Stop node processing process for entire layer upon encountering a node which is not ready for computation. This guaranties the correct order of computation, but would slow down the loading process to a certain degree.
2. Note the original node order while iterating over them and fill the node_index_map keys accordingly. This way the original order will be broken, but we would have information about it while searching for the following nodes' inputs.

**Standalone code to reproduce the issue**
https://colab.research.google.com/drive/1F8Rii8SmlURIlBK7GmMIkJwt5YkT6E7F?usp=sharing

**Comment by UsharaniPagadala** [comment]
@Saduf2019 
I was able to replicate the issue reported here.Please find the [gist](https://colab.research.google.com/gist/UsharaniPagadala/1244d3ced00934198e3f28e624f1c0cb/keras_bug_example.ipynb#scrollTo=MnhrUXsMuFap). Thanks

**Comment by ymodak** [comment]
Please post this issue on [keras-team/keras](https://github.com/keras-team/keras/issues) repo.
To know more see;
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999

**Comment by hidanom** [comment]
> Please post this issue on [keras-team/keras](https://github.com/keras-team/keras/issues) repo.
> To know more see;
> https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999

I've [duplicated](https://github.com/keras-team/keras/issues/14838) the issue. Hope it helps.

Could you maybe provide an estimation on when someone is going to fix the problem? It seems to be a critical one, as h5 saving format is essentially broken.

**Comment by hidanom** [comment]
This issue should not be stale. I have no idea why "awaiting response" label is still on.

**Comment by bhack** [comment]
Isn't tracked now at https://github.com/keras-team/keras/issues/14838?

**Comment by jvishnuvardhan** [comment]
@hidanom closing this issue as this was resolved. Thanks for confirming here https://github.com/keras-team/keras/issues/14838
